### ISSUE TITLE
Require at least Ubuntu 18.04

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#3664]
OpenTitan currently uses Ubuntu 16.04 as oldest supported version of Ubuntu, but also as baseline for other distributions ("if they aren't older than Ubuntu 16.04, things should be fine"). This issue suggests moving the baseline to Ubuntu 18.04 LTS. 

Motivation for going from Ubuntu 16.04 to Ubuntu 18.04

* Ubuntu 16.04 (as the name implies) has been released in April 2016 and will reach its end of life in April 2021 (https://help.ubuntu.com/community/EOL#Ubuntu_16.04_Xenial_Xerus; extended support releases might be available after that)
* Primary motivation: Ubuntu 16.04 comes with Python 3.5, which "will" be [EOL in September 2020](https://www.python.org/downloads/release/python-3510/). Even though the Python package in Ubuntu will still receive security fixes, more and more third-party Python packages are dropping support for Python 3.5.
  * [Meson 0.57 will require Python 3.6](https://github.com/mesonbuild/meson/blob/master/docs/markdown/snippets/python35-eol.md)
  * pip (the python package installer) will drop Python 3.5-suppport soon ("DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.")
  * Python 3.6 provides some features which many developers have already been trying to use, and we had to go back and remove them: Pathlib in Python 3.6 is significantly easier to use, since Path objects are now accepted in calls like `open()`. f strings are supported. etc.
* Clang(-format) will be available in version 6.0 (vs. 3.8 in Ubuntu 16.04)
* (Host) GCC moves from 5.4 to 7.4.0

Discussion: I expect the transition to be smooth and mostly unnoticed by users.

* As so often, EDA tool support is a major reason to stay with older OS releases. However, Ubuntu 18.04 is already "old enough" so that existing workflows are expected to continue to work.
  * Vivado 2018.3+ supports Ubuntu 18.04. That means we do have a couple of old Vivado versions available if necessary.
  * Synopsys products in the P (2019) and Q (2020) platform releases claim to support "Debian 9.0" (released mid-2017) and have never officially supported Ubuntu. However, we know that VCS works well with Ubuntu 18.04, DC also typically works well. 
  * Cadence tools and more specialized design tools are often more tricky to get working under Ubuntu in the first place, and many of them require RHEL/CentOS. The Ubuntu version change shouldn't affect those users, as long as they are able to obtain a sufficiently recent Python version (e.g. from a RHEL SCL). This is no different from today.
  * None of the open source EDA tools we're using have any problem with newer Ubuntu versions.
* Many users (both on the Google side, and here at lowRISC) have been using a variety of newer distributions, including Ubuntu 18.04, for work for a long time now and have not seen issues.

How to do it:
1. Form consensus around this move.
2. Announce it publicly if agreed.
3. Update the documentation, and switch all CI base images to Ubuntu 18.04

Please comment if you have seen problems or success stories with Ubuntu 18.04, or have seen issues where components of Ubuntu 18.04 are too new for a certain tool to work.

**Comment by gkelly** [comment]
I'm strongly in favor of this. We also know that the docs builder can run just fine on 18.04 (it had been doing that for quite a while before we downgraded it to match the required Ubuntu version).

**Comment by eunchan** [comment]
+1 to move to newer version. not using `f""` syntax and `dict` as OrderedDict was a little bit of pain to me. :) When we move to newer version, which python version do you think we should aim? I think each minor version of Python3 is being deprecated every year. Should we aim for 3.8 to be save for 3 years? :)

**Comment by vogelpi** [comment]
Thanks @imphil for starting this discussion. I am in favor of moving to 18.04, too. I have been using this distribution for 1.5 years now and didn't experience issues.

@eunchan I think the idea is to keep using Python 3.6 for now. Switching the Python version is probably a different discussion.

**Comment by imphil** [comment]
@eunchan I'd love to require Python 3.8 right away, but using a version of Python which isn't the default Python version of the distribution you're using is always a bit of a pain. That would give us Python 3.6 if we go with Ubuntu 18.04, which I think it a good step forward and should address the main pain points we have at the moment with Python 3.5.

**Comment by eunchan** [comment]
Yeah, Python 3.6 is also fine :)

**Comment by imphil** [comment]
@rasmus-madsen @martin-lueker @zi-v it would be good to get an "OK" from your side to move ahead with this. Please let me know if you're unclear about the implications on your side.

**Comment by martin-lueker** [comment]
Hi @imphil,  I'll leave it to @rasmus-madsen to confirm, but I fear that this would be an absolute disaster from the WDC side, as our server farm is persisting on using RHEL6.   Now I can't claim to know what leaving Ubuntu 16.04 behind would mean, for us stranded on RHEL6.  Is the idea to keep the existing package requirements, but simply stop 16.04 testing?  (That in itself could lead to unexpected problems for us).

Would there be any new packages as a result of this change?

In short, it makes me nervous until I have confirmation that our Sysadmins are willing to leave the Stone Age.

**Comment by zi-v** [comment]
Nuvoton are OK. We can install Ubuntu 18.04 if needed.

**Comment by rasmus-madsen** [comment]
@imphil unfortunately @martin-lueker is correct in his statement.

I will try to bring it up with our sysadmins again and see if they are any closer to a solution.

**Comment by imphil** [comment]
@rbohn can you check or put us in contact with someone on your side if this change would be OK?

**Comment by rbohn** [comment]
Some EDA tool background:

There is an industry group that tries to keep the various EDA vendors in sync in terms of OS support. They have some guidelines here (http://esd-alliance.org/interoperability-committee/) which is basically RHEL 7 now and beginning of RHEL 8 transition next year.

Just recently some EDA vendors announced dropping support for RHEL 6 and moving the future baseline to RHEL 7. e.g. https://www.cadence.com/content/dam/cadence-www/global/en_US/documents/support/RHEL_6_EOL.pdf . Production EDA environments can be slow to move because of the length of time that ASIC projects need to be supported. 

For the moment, the standard for new project EDA environments is RHEL 7. With that in mind, note that the version of Ubuntu you are seeking to deprecate is 2 years newer than the initial release of RHEL 7.

That said, I'm not sure this would really be an issue. It looks like one of the problems you are trying to solve is access to Python 3.6. We already have this and newer versions of python available even on our CentOS 6 era machines. The biggest compatibility problem I've encountered on older machines is the GLIBC 2.14 memcpy issue, which is addressed with anything post CentOS 6 I believe. I've personally built for example GCC 10.2 on CentOS 7 machines without issue and haven't generally experienced any fundamental compatibility problems when building newer packages from source on those machines, which we will do when necessary.

**Comment by imphil** [comment]
@rbohn thanks for your input, sounds good.

From what I've heard here and on other channels we are in agreement that we can move forward with this proposal. The TC is also aware of this change and decided to let it move forward without a formal RFC.

If anybody finds compat issues with their setup after we make the switch please let me know, and I'm reasonably certain that we'll find a good solution.

**Comment by martin-lueker** [comment]
@imphil,
   Perhaps we can refine the language on this somehow?   Make a page "Checking your distribution for OT compatibility", perhaps?

**Comment by martin-lueker** [comment]
I say this because jumping to different hosts for different applications is going to be a nightmare once the impacts make it mandatory for us to everything but sims on Ubuntu 18.4 or similar.

**Comment by rasmus-madsen** [comment]
> I say this because jumping to different hosts for different applications is going to be a nightmare once the impacts make it mandatory for us to everything but sims on Ubuntu 18.4 or similar.

@imphil I think martin is on to something here.
since we can't expect everyone to run Ubuntu,
it might be a good idea to elaborate some more on what the requirements are for 
Python
Glibc
Gcc
ninja 
etc

also we could add a known issues section with older distributions  - like the Glibc issue for RHEL6
